src/almond/worker_k8s.ts

```diff
@@ -0,0 +1,304 @@
+#!/usr/bin/env node
+// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// This file is part of Almond
+//
+// Copyright 2021 The Board of Trustees of the Leland Stanford Junior University
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// load thingpedia to initialize the polyfill
+import 'thingpedia';
+
+import express from 'express';
+import WebSocket from "ws";
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const wsjs = require('ws');
+
+import * as http from 'http';
+import * as stream from 'stream';
+import * as rpc from 'transparent-rpc';
+import * as argparse from 'argparse';
+
+import PlatformModule, { PlatformOptions } from './platform';
+import JsonDatagramSocket from '../util/json_datagram_socket';
+import * as i18n from '../util/i18n';
+import Engine from './engine';
+
+interface EngineState {
+    userId : number;
+    running : boolean;
+    sockets : Set<rpc.Socket>;
+    stopped : boolean;
+    engine ?: Engine;
+}
+
+class Worker {
+    private app : express.Application;
+    private server : http.Server;
+    private wss : WebSocket.Server;
+    private engines : Map<number, EngineState>; 
+    private stopped : boolean;
+
+    constructor(port : number) {
+        this.engines = new Map<number, EngineState>();
+        this.stopped = false;
+        this.app = express();
+        this.app.use(express.json());
+        this.app.use(express.urlencoded({ extended: true }));
+        this.app.set('port', port);
+
+        this.server = http.createServer(this.app);
+        this.server.on('upgrade', (request, socket, head) => {
```

Comment:
You might want to use `expressWs` instead, like we do in the frontend. Then you can handle websockets as normal routes with `app.ws(path, callback)`

Reply:
Please see my latest commit. I think I need to hijack the `upgrade` to create the webstream. Let me know if it makes sense.

Reply:
This is still not right to me. If you use `expressWs`you get the same `ws` object that you have in the callback to `handleUpgrade`. This bit independent of what we do about transparent-rpc.